pkg/kp/consulutil/watch.go

```diff
@@ -0,0 +1,165 @@
+		case <-timer.C:
+		}
+		timer.Reset(250 * time.Millisecond) // upper bound on request rate
+		pairs, queryMeta, err := SafeList(clientKV, done, prefix, &api.QueryOptions{
+			WaitIndex: currentIndex,
+		})
+		switch err {
+		case CanceledError:
+			return
+		case nil:
+			currentIndex = queryMeta.LastIndex
+			select {
+			case <-done:
+			case outPairs <- pairs:
+			}
+		default:
+			select {
+			case <-done:
+			case outErrors <- err:
+			}
+			timer.Reset(2 * time.Second) // backoff
+		}
+	}
+}
+
+type NewKeyHandler func(key string) chan<- *api.KVPair
+
+type keyMeta struct {
+	created    uint64
+	modified   uint64
+	subscriber chan<- *api.KVPair
+}
+
+// WatchNewKeys watches for changes to a list of Key/Value pairs and lets each key be
+// handled individually though a subscription-like interface.
+//
+// This function models a key's lifetime in the following way. When a key is first seen,
+// the given NewKeyHandler function will be run, which may return a channel. When the
+// key's value changes, new K/V updates are sent to the key's notification channel. When
+// the key is deleted, `nil` is sent. After being deleted or if the watcher is asked to
+// exit, a key's channel will be closed, to notify the receiver that no further updates
+// are coming.
+//
+// WatchNewKeys doesn't watch a prefix itself--the caller should arrange a suitable input
+// stream of K/V pairs, probably from WatchPrefix(). This function runs until the input
+// stream closes. Closing "done" will asynchronously cancel the watch and cause it to
+// eventually exit.
+func WatchNewKeys(pairsChan <-chan api.KVPairs, onNewKey NewKeyHandler, done <-chan struct{}) {
+	keys := make(map[string]*keyMeta)
+
+	defer func() {
+		for _, keyMeta := range keys {
+			if keyMeta.subscriber != nil {
+				close(keyMeta.subscriber)
+			}
+		}
+	}()
+
+	for {
+		var pairs api.KVPairs
+		var ok bool
+		select {
+		case <-done:
+			return
+		case pairs, ok = <-pairsChan:
+			if !ok {
+				return
+			}
+		}
+
+		visited := make(map[string]bool)
+
+		// Scan for new and changed keys
+		for _, pair := range pairs {
+			visited[pair.Key] = true
+			if keyMeta, ok := keys[pair.Key]; ok {
+				if keyMeta.created == pair.CreateIndex {
+					// Existing key that was seen before
+					if keyMeta.subscriber != nil && keyMeta.modified != pair.ModifyIndex {
+						// It's changed!
+						keyMeta.modified = pair.ModifyIndex
+						select {
+						case <-done:
+							return
+						case keyMeta.subscriber <- pair:
+						}
+					}
+					continue
+				} else {
+					// This key was deleted and recreated between queries
+					if keyMeta.subscriber != nil {
+						select {
+						case <-done:
+							return
+						case keyMeta.subscriber <- nil:
+						}
+						close(keyMeta.subscriber)
+					}
+					// Fall through to re-create this key
+				}
+			}
+			// Found a new key
+			keyMeta := &keyMeta{
+				created:    pair.CreateIndex,
+				modified:   pair.ModifyIndex,
+				subscriber: onNewKey(pair.Key),
```

Comment:
My first versions of this method did just that.  It doesn't interact well with middle layers that have to parse these raw values into concrete types and might fail.

Imagine that someone is manually modifying, say, resource controller JSON definitions and creates a key that doesn't parse correctly.  If a client picks that instance to start watching the rcstore, what will the store deliver to the client?  Delay the new key notification until a clean value arrives?  Send the notification without a value?  The former adds a lot of complexity to the _type_-store, the latter adds complexity to the consumers.

There is less total complexity when the new key notification is separate from the value stream itself.  And there won't be that much blocking going on: sending to the returned channel happens a few lines later.